### PR TITLE
Added insn overlapping (bitmanip / krypto)

### PR DESCRIPTION
**Pull request by marcfedorow**

### CODE REVIEW

**Comment by marcfedorow** [comment]
@aswaterman @ben-marshall please review

File: riscv/insns/gorci.h [hunk 1]
```diff
@@ -1,8 +1,12 @@
...
-  require_extension(EXT_ZBB);
-else
+try {
+  if (SHAMT == 7) {
+    require_extension(EXT_ZBB);
+  }
+  else throw trap_illegal_instruction(0);
+} catch(...) {
   require_extension(EXT_XBITMANIP);
+}
```
Comment (aswaterman) [inline riscv/insns/gorci.h]: This looks correct but rather convoluted.  I think we want:

    require((p->extension_enabled(EXT_ZBB) && SHAMT == 7) || p->extension_enabled(EXT_XBITMANIP));

File: riscv/insns/pack.h [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
-if (insn.rs2() == 0 && xlen == 32)
+try {
+  require_either_extension(EXT_ZBKB, EXT_XBITMANIP);
+} catch (...) {
+  require(insn.rs2() == 0);
+  require_rv32;
   require_extension(EXT_ZBB);
-else
-  require_extension(EXT_XBITMANIP);
+}
```
Comment (aswaterman) [inline riscv/insns/pack.h]: See comment on `gorci`

File: riscv/insns/packw.h [hunk 3]
```diff
@@ -1,8 +1,10 @@
 // RV64Zbb contains zext.h but not general packw
-if (insn.rs2() == 0)
+try {
+  require_either_extension(EXT_ZBKB, EXT_XBITMANIP);
+} catch(...) {
+  require(insn.rs2() == 0);
   require_extension(EXT_ZBB);
-else
-  require_extension(EXT_XBITMANIP);
+}
```
Comment (aswaterman) [inline riscv/insns/packw.h]: See comment on `gorci`

File: riscv/insns/shfli.h [hunk 4]
```diff
@@ -1,4 +1,10 @@
-require_extension(EXT_XBITMANIP);
+// Zbkb contains zip but not general packw
+try {
+  require(insn.rs2() == (xlen / 2 - 1));
+  require_extension(EXT_ZBKB);
+} catch (...) {
+  require_extension(EXT_XBITMANIP);
+}
```
Comment (aswaterman) [inline riscv/insns/shfli.h]: See comment on `gorci`

File: riscv/insns/unshfli.h [hunk 5]
```diff
@@ -1,4 +1,10 @@
-require_extension(EXT_XBITMANIP);
+// Zbkb contains unzip but not general packw
+try {
+  require(insn.rs2() == (xlen / 2 - 1));
+  require_extension(EXT_ZBKB);
+} catch (...) {
+  require_extension(EXT_XBITMANIP);
+}
```
Comment (aswaterman) [inline riscv/insns/unshfli.h]: See comment on `gorci`

**Comment by ben-marshall** [comment]
Apart from @aswaterman's comments this looks good to me :+1:

**Comment by marcfedorow** [comment]
@aswaterman done

**Comment by aswaterman** [comment]
Looks good from my perspective!

**Comment by scottj97** [comment]
I'm seeing some new compiler warnings from this PR:
```
g++ -MMD -MP -I/include/boost-0 -DPREFIX=\"/usr/local\" -Wall -Wno-unused -g -O2 -fPIC -std=c++11 -g -O2   -I. -I/my/path/riscv-isa-sim -I/my/path/riscv-isa-sim/fesvr -I/my/path/riscv-isa-sim/riscv -I/my/path/riscv-isa-sim/disasm -I/my/path/riscv-isa-sim/customext -I/my/path/riscv-isa-sim/fdt -I/my/path/riscv-isa-sim/softfloat -I/my/path/riscv-isa-sim/spike_main -I/my/path/riscv-isa-sim/spike_dasm   -c shfli.cc
In file included from /my/path/riscv-isa-sim/riscv/decode.h:16,
                 from /my/path/riscv-isa-sim/riscv/mmu.h:6,
                 from /my/path/riscv-isa-sim/riscv/insn_template.h:4:
/my/path/riscv-isa-sim/riscv/insns/shfli.h: In function ‘reg_t rv32_shfli(processor_t*, insn_t, reg_t)’:
/my/path/riscv-isa-sim/riscv/insns/shfli.h:2:22: warning: comparison of integer expressions of different signedness: ‘uint64_t’ {aka ‘long unsigned int’} and ‘int’ [-Wsign-compare]
 require(((insn.rs2() == (xlen / 2 - 1)) && p->extension_enabled(EXT_ZBKB))
           ~~~~~~~~~~~^~~~~~~~~~~~~~~~~
/my/path/riscv-isa-sim/riscv/common.h:8:39: note: in definition of macro ‘unlikely’
 # define unlikely(x) __builtin_expect(x, 0)
                                       ^
/my/path/riscv-isa-sim/riscv/insns/shfli.h:2:1: note: in expansion of macro ‘require’
 require(((insn.rs2() == (xlen / 2 - 1)) && p->extension_enabled(EXT_ZBKB))
 ^~~~~~~
/my/path/riscv-isa-sim/riscv/insns/shfli.h: In function ‘reg_t rv64_shfli(processor_t*, insn_t, reg_t)’:
/my/path/riscv-isa-sim/riscv/insns/shfli.h:2:22: warning: comparison of integer expressions of different signedness: ‘uint64_t’ {aka ‘long unsigned int’} and ‘int’ [-Wsign-compare]
 require(((insn.rs2() == (xlen / 2 - 1)) && p->extension_enabled(EXT_ZBKB))
           ~~~~~~~~~~~^~~~~~~~~~~~~~~~~
/my/path/riscv-isa-sim/riscv/common.h:8:39: note: in definition of macro ‘unlikely’
 # define unlikely(x) __builtin_expect(x, 0)
                                       ^
/my/path/riscv-isa-sim/riscv/insns/shfli.h:2:1: note: in expansion of macro ‘require’
 require(((insn.rs2() == (xlen / 2 - 1)) && p->extension_enabled(EXT_ZBKB))
 ^~~~~~~
```

**Comment by aswaterman** [comment]
Fixed.

**Pull request merged**